bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/BuildManager.java

```diff
@@ -1272,22 +1273,48 @@ private boolean needsBuild(InternalBuilder builder, int trigger) {
 		});
 
 		//search for the builder's project
-		if (currentDelta.findNodeAt(builder.getProject().getFullPath()) != null) {
-			if (Policy.DEBUG_BUILD_NEEDED)
-				Policy.debug(toString(builder) + " needs building because of changes in: " + builder.getProject().getName()); //$NON-NLS-1$
-			return true;
+		boolean ret = false;
+		AbstractDataTreeNode found;
+		if ((found = currentDelta.findNodeAt(builder.getProject().getFullPath())) != null) {
+			if (Policy.DEBUG_BUILD_NEEDED) {
+				String prefix = toString(builder);
+				Policy.debug(prefix + " needs building because of changes in: " + builder.getProject().getName()); //$NON-NLS-1$
+				if (debugBuildNeededDelta())
+					debugDeltaNode(prefix, found);
+			}
+			ret = true;
+			if (!debugBuildNeededDelta())
```

Comment:
bad that the control flow is changed by debugging option. Could that be resolved by copy&paste some code from below?

Reply:
The control flow by itself does require to find the _first_ entry matching either Project or InterestingProjects (without need to find all of them), but when debugging I want to see _each_ matching entry.